### PR TITLE
v0.8.2 : add data source logic and link it to variable, remove CalculatedVariable

### PR DESCRIPTION
**Pull request by clavay**
Minor changes:
- add `operations` and `aggregation` protocols
- fix migration `WidgetCotent` update for `Charts`
- plugin can add apps to `INSTALLED_APPS` in `settings.py`
- fix for `get_objects_for_html`

Major changes:
- add `DataSourceModel` (define how a data source appears in the admin), `Datasource` (base for all datasource) and `DjangoDatabase` (allows to define data sources as django models like `RecordedData` or `RecordedDataOld`)
- create `VariableManager` (replacing the `RecordedDataValueManager` to support data sources)
- delete `CalculatedVariable` : moved to the [PyScada-Operations](https://github.com/clavay/PyScada-Operations) plugin (aggregation protocol)

### CODE REVIEW

File: docs/device_protocol.rst [hunk 1]
```diff
@@ -19,6 +19,8 @@ Device Protocol IDs
 - 15: `MeterBus (MBus) <https://github.com/pyscada/PyScada-MeterBus>`_
 - 16: Generic dummy device
 - 17: `EMS <https://github.com/pyscada/PyScada-EMS>`_
+- 18: `Operations <https://github.com/pyscada/PyScada-Operations>`_
+- 19: `Aggregation <https://github.com/pyscada/PyScada-Operations>`_
```
Comment (trombastic) [inline docs/device_protocol.rst:23]: the URL seems to be wrong

**Review by trombastic — COMMENTED** [review]
do the changes of this commit allow to keep existing data in the "old" RecordedData Database?

File: pyscada/apps.py [hunk 2]
```diff
@@ -29,3 +29,61 @@ def ready(self):
...
+            )
+            dd, _ = DjangoDatabase.objects.get_or_create(
+                datasource=ds,
+                defaults={
+                    "data_model_app_name": "pyscada",
+                    "data_model_name": "RecordedDataOld",
+                },
+            )
+
+        except ProgrammingError:
```
[the remark this answers is not in the document]
Comment (clavay) [reply pyscada/apps.py:86]: `ProgrammingError` and `OperationalError` appears when migrations are not yet applied.
Do you think we should add log for this case ?

File: docs/device_protocol.rst (hunk 1 above)
Comment (clavay) [reply docs/device_protocol.rst:23]: It is in my [account](https://github.com/pyscada), so the pyscada organization can fork it to make to link exist.
what do you think ?

**Comment by clavay** [comment]
> do the changes of this commit allow to keep existing data in the "old" RecordedData Database?

Are you talking about the `RecordedDataOld` table ?
The changes does not touch the RecordedDataOld.
But we need to specify the manager to use for this model, for this we can adapt the RecordedDataManager to never use `date_saved`.
What do you think ?

File: docs/device_protocol.rst (hunk 1 above)
Comment (trombastic) [reply docs/device_protocol.rst:23]: okay, fork is done

both apps have the same Repo so the URL is okay

File: pyscada/apps.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (trombastic) [reply pyscada/apps.py:86]: yes i think so, otherwise it is not visible for the user/admin what happend

**Comment by trombastic** [comment]
if i have a working pyscada installation with data in the RecordedData table, will this data be preserved and accessible after the update?

**Comment by clavay** [comment]
Yes, default ([here](https://github.com/clavay/PyScada/blob/WriteReadRecordedDataCustomFcts_CP/pyscada/models.py#L1747)) is for each variable to use the first (`id=1`, [here](https://github.com/clavay/PyScada/blob/WriteReadRecordedDataCustomFcts_CP/pyscada/apps.py#L49)) data source, which is the `DataSource` linked to the `RecordedData` table.

A plugin can add datasource using the [DjangoDatabase model](https://github.com/clavay/PyScada/blob/WriteReadRecordedDataCustomFcts_CP/pyscada/models.py#L1520-L1654) and define another table to use (as [here](https://github.com/clavay/PyScada/blob/WriteReadRecordedDataCustomFcts_CP/pyscada/apps.py#L62-L84) for the `RecordedDataOld` table).

A plugin can also, as the `PyScada-Operations` plugin, add another [DataSourceModel](https://github.com/clavay/PyScada-Operations/blob/main/pyscada/operations/apps.py#L25-L33), create a [DataSource](https://github.com/clavay/PyScada-Operations/blob/main/pyscada/operations/apps.py#L35) and [create a model](https://github.com/clavay/PyScada-Operations/blob/main/pyscada/operations/apps.py#L37) with a OneToOne relation to the `DataSource` model, specifying additional fields and the functions needed (read, write, last element... as described in the commit 7110864 message), as done by the [OperationsDataSource](https://github.com/clavay/PyScada-Operations/blob/main/pyscada/operations/models.py#L70-L426)

I will publish a simple plugin creating random data in the read function and logging values in the write function as an example.
I will also add some documentation later.

File: pyscada/apps.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clavay) [reply pyscada/apps.py:86]: I will do it on Monday.

**Comment by clavay** [comment]
I add the logs for [ProgrammingError and OperationalError while populating models](https://github.com/pyscada/PyScada/pull/109/commits/1e4879ce416ca81fdae22d120f74bd9d64ba0752)

Is it ok for merging ?

**Review by trombastic — APPROVED** [review]

**Pull request merged**